drivers/serial/uart_ite_it8xxx2.c

```diff
@@ -0,0 +1,147 @@
+
+#include <device.h>
+#include <drivers/gpio.h>
+#include <drivers/uart.h>
+#include <kernel.h>
+#include <pm/device.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(uart_ite_it8xxx2, LOG_LEVEL_ERR);
+
+struct uart_it8xxx2_config {
+	uint8_t port;
+	uint8_t wakeup_pin;
+	/* GPIO handle */
+	const struct device *gpio_dev;
+	/* UART handle */
+	const struct device *uart_dev;
+};
+
+enum uart_port_num {
+	UART1 = 1,
+	UART2,
+};
+
+#ifdef CONFIG_PM_DEVICE
+__weak void uart1_wui_isr_late(void) { }
+__weak void uart2_wui_isr_late(void) { }
+
+void uart1_wui_isr(const struct device *gpio, struct gpio_callback *cb,
+		   uint32_t pins)
+{
+	/* Disable interrupts on UART1 RX pin to avoid repeated interrupts. */
+	gpio_pin_interrupt_configure(gpio, (find_msb_set(pins) - 1),
+				     GPIO_INT_DISABLE);
+
+	if (uart1_wui_isr_late) {
+		uart1_wui_isr_late();
+	}
+}
+
+void uart2_wui_isr(const struct device *gpio, struct gpio_callback *cb,
+		   uint32_t pins)
+{
+	/* Disable interrupts on UART2 RX pin to avoid repeated interrupts. */
+	gpio_pin_interrupt_configure(gpio, (find_msb_set(pins) - 1),
+				     GPIO_INT_DISABLE);
+
+	if (uart2_wui_isr_late) {
+		uart2_wui_isr_late();
+	}
+}
+
+static inline int uart_it8xxx2_pm_action_cb(const struct device *dev,
+					    enum pm_device_action action)
+{
+	const struct uart_it8xxx2_config *const config = dev->config;
+	int ret = 0;
+
+	switch (action) {
+	/* Next device power state is in active. */
+	case PM_DEVICE_ACTION_RESUME:
+		/* To do when EC is resuming. */
+		break;
+	/* Next device power state is deep doze mode */
+	case PM_DEVICE_ACTION_SUSPEND:
+		/* Enable UART WUI */
+		ret = gpio_pin_interrupt_configure(config->gpio_dev,
+						   config->wakeup_pin,
+						   GPIO_INT_TRIG_LOW);
+		if (ret) {
+			LOG_ERR("Failed to configure UART%d WUI (ret %d)",
+				config->port, ret);
+			return ret;
+		}
+		/*
+		 * If uart device is busy with transmitting, the driver will
+		 * stay in while loop and wait for the transaction is completed.
+		 */
+		while (!uart_irq_tx_complete(config->uart_dev)) {
+			continue;
+		}
+
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+#endif /* CONFIG_PM_DEVICE */
+
+static int uart_it8xxx2_init(const struct device *dev)
+{
+	__unused const struct uart_it8xxx2_config *const config = dev->config;
+
+#ifdef CONFIG_PM_DEVICE
+	int ret = 0;
+
+	if (config->port == UART1) {
+		static struct gpio_callback uart1_wui_cb;
+
+		gpio_init_callback(&uart1_wui_cb, uart1_wui_isr,
+				   BIT(config->wakeup_pin));
+
+		ret = gpio_add_callback(config->gpio_dev, &uart1_wui_cb);
+	} else if (config->port == UART2) {
+		static struct gpio_callback uart2_wui_cb;
+
+		gpio_init_callback(&uart2_wui_cb, uart2_wui_isr,
+				   BIT(config->wakeup_pin));
+
+		ret = gpio_add_callback(config->gpio_dev, &uart2_wui_cb);
+	}
+
+	if (ret) {
+		LOG_ERR("Failed to add UART%d callback (err %d)",
+			config->port, ret);
+		return ret;
+	}
```

Comment:
this looks like a gpio based wakeup, not really a wakeup based on e.g. uart activity, so I wonder if this can be made more generic

Reply:
Add a comment to explain that wakeup source is through UART Rx interrupt.